src/framework/auth/strategies/oauth2/oauth2-strategy.ts

```diff
@@ -0,0 +1,321 @@
+import { switchMap } from 'rxjs/operators/switchMap';
+import { map } from 'rxjs/operators/map';
+import { catchError } from 'rxjs/operators/catchError';
+import { NB_WINDOW } from '@nebular/theme';
+
+import { NbAuthStrategy } from '../auth-strategy';
+import { NbAuthRefreshableToken, NbAuthResult } from '../../services/';
+import { NbOAuth2AuthStrategyOptions, NbOAuth2ResponseType, auth2StrategyOptions } from './oauth2-strategy.options';
+
+
+/**
+ * OAuth2 authentication strategy.
+ *
+ * @example
+ *
+ * Strategy settings:
+ *
+ * ```
+ *
+ * export enum NbOAuth2ResponseType {
+ *   CODE = 'code',
+ *   TOKEN = 'token',
+ * }
+ *
+ * // TODO: password, client_credentials
+ * export enum NbOAuth2GrantType {
+ *   AUTHORIZATION_CODE = 'authorization_code',
+ *   REFRESH_TOKEN = 'refresh_token',
+ * }
+ *
+ * export class NbOAuth2AuthStrategyOptions {
+ *   name: string;
+ *   baseEndpoint?: string = '';
+ *   clientId: string = '';
+ *   clientSecret: string = '';
+ *   redirect?: { success?: string; failure?: string } = {
+ *     success: '/',
+ *     failure: null,
+ *   };
+ *   defaultErrors?: any[] = ['Something went wrong, please try again.'];
+ *   defaultMessages?: any[] = ['You have been successfully authenticated.'];
+ *   authorize?: {
+ *     endpoint?: string;
+ *     redirectUri?: string;
+ *     responseType?: string;
+ *     scope?: string;
+ *     state?: string;
+ *     params?: { [key: string]: string };
+ *   } = {
+ *     endpoint: 'authorize',
+ *     responseType: NbOAuth2ResponseType.CODE,
+ *   };
+ *   token?: {
+ *     endpoint?: string;
+ *     grantType?: string;
+ *     redirectUri?: string;
+ *     class: NbAuthTokenClass,
+ *   } = {
+ *     endpoint: 'token',
+ *     grantType: NbOAuth2GrantType.AUTHORIZATION_CODE,
+ *     class: NbAuthOAuth2Token,
+ *   };
+ *   refresh?: {
+ *     endpoint?: string;
+ *     grantType?: string;
+ *     scope?: string;
+ *   } = {
+ *     endpoint: 'token',
+ *     grantType: NbOAuth2GrantType.REFRESH_TOKEN,
+ *   };
+ * }
+ * ```
+ *
+ */
+@Injectable()
+export class NbOAuth2AuthStrategy extends NbAuthStrategy {
+
+  get isTypeCode(): boolean {
+    return this.getOption('authorize.responseType') === NbOAuth2ResponseType.CODE;
+  }
+
+  get isTypeToken(): boolean {
+    return this.getOption('authorize.responseType') === NbOAuth2ResponseType.TOKEN;
+  }
+
+  protected defaultOptions: NbOAuth2AuthStrategyOptions = auth2StrategyOptions;
+
+  constructor(protected http: HttpClient,
+              private route: ActivatedRoute,
+              @Inject(NB_WINDOW) private window: any) {
+    super();
+  }
+
+  authenticate(): Observable<NbAuthResult> {
+    return this.isRedirectResult()
+      .pipe(
+        switchMap((result: boolean) => {
+          if (!result) {
+            this.authorizeRedirect();
+            return observableOf(null);
+          }
+          return this.getAuthorizationResult();
+        }),
+      );
+  }
+
+  getAuthorizationResult(): Observable<any> {
```

Comment:
As I understand correctly there will be more types: code, token, password, credentials. So, let's create a map with type -> redirect result instead of *isType** functions and ugly if. What do you think?